File: lib/rayz/chapter13.rb
module Rayz
  class Chapter13
    def self.run
      puts "\n=== Chapter 13: Groups ==="
      puts "Demonstrating hierarchical scene composition with groups..."

      # Create camera
      camera = Camera.new(
        hsize: 800,
        vsize: 600,
        field_of_view: Math::PI / 3
      )
      camera.transform = Transformations.view_transform(
        from: Point.new(x: 0, y: 3, z: -8),
        to: Point.new(x: 0, y: 1, z: 0),
        up: Vector.new(x: 0, y: 1, z: 0)
      )

      # Create light source
      light = PointLight.new(
        position: Point.new(x: -5, y: 10, z: -10),
        intensity: Color.new(r: 1, g: 1, b: 1)
      )

      # Create floor
      floor = Plane.new
      floor.material.pattern = CheckersPattern.new(
        Color.new(r: 0.9, g: 0.9, b: 0.9),
        Color.new(r: 0.1, g: 0.1, b: 0.1)
      )
      floor.material.reflective = 0.1

      # Create a simple "tree" using groups
      # Trunk group (cylinder)
      trunk = Cylinder.new
      trunk.minimum = 0
      trunk.maximum = 1.5
      trunk.closed = true
      trunk.material.color = Color.new(r: 0.4, g: 0.2, b: 0.1)
      trunk.material.specular = 0.1

      # Foliage group (multiple spheres)
      foliage_group = Group.new

      # Center foliage
      foliage1 = Sphere.new
      foliage1.transform = Transformations.translation(x: 0, y: 2, z: 0) *
        Transformations.scaling(x: 0.8, y: 0.8, z: 0.8)
      foliage1.material.color = Color.new(r: 0.1, g: 0.6, b: 0.1)
      foliage1.material.specular = 0.3

      # Left foliage
      foliage2 = Sphere.new
      foliage2.transform = Transformations.translation(x: -0.5, y: 1.7, z: 0) *
        Transformations.scaling(x: 0.6, y: 0.6, z: 0.6)
      foliage2.material.color = Color.new(r: 0.1, g: 0.7, b: 0.1)
      foliage2.material.specular = 0.3

      # Right foliage
      foliage3 = Sphere.new
      foliage3.transform = Transformations.translation(x: 0.5, y: 1.7, z: 0.2) *
        Transformations.scaling(x: 0.6, y: 0.6, z: 0.6)
      foliage3.material.color = Color.new(r: 0.1, g: 0.8, b: 0.1)
      foliage3.material.specular = 0.3

      # Top foliage
      foliage4 = Sphere.new
      foliage4.transform = Transformations.translation(x: 0, y: 2.5, z: 0) *
        Transformations.scaling(x: 0.5, y: 0.5, z: 0.5)
      foliage4.material.color = Color.new(r: 0.1, g: 0.5, b: 0.1)
      foliage4.material.specular = 0.3

      foliage_group.add_child(foliage1)
      foliage_group.add_child(foliage2)
      foliage_group.add_child(foliage3)
      foliage_group.add_child(foliage4)

      # Tree group (trunk + foliage)
      tree = Group.new
      tree.add_child(trunk)
      tree.add_child(foliage_group)
      tree.transform = Transformations.translation(x: -2, y: 0, z: 1)

      # Create a "snowman" using groups
      # Bottom sphere
      snowman_bottom = Sphere.new
      snowman_bottom.transform = Transformations.translation(x: 0, y: 0.6, z: 0) *
        Transformations.scaling(x: 0.6, y: 0.6, z: 0.6)
      snowman_bottom.material.color = Color.new(r: 1, g: 1, b: 1)
      snowman_bottom.material.specular = 0.9
      snowman_bottom.material.shininess = 300

      # Middle sphere
      snowman_middle = Sphere.new
      snowman_middle.transform = Transformations.translation(x: 0, y: 1.5, z: 0) *
        Transformations.scaling(x: 0.45, y: 0.45, z: 0.45)
      snowman_middle.material.color = Color.new(r: 1, g: 1, b: 1)
      snowman_middle.material.specular = 0.9
      snowman_middle.material.shininess = 300

      # Head sphere
      snowman_head = Sphere.new
      snowman_head.transform = Transformations.translation(x: 0, y: 2.2, z: 0) *
        Transformations.scaling(x: 0.3, y: 0.3, z: 0.3)
      snowman_head.material.color = Color.new(r: 1, g: 1, b: 1)
      snowman_head.material.specular = 0.9
      snowman_head.material.shininess = 300

      # Nose (carrot)
      nose = Cylinder.new
      nose.minimum = 0
      nose.maximum = 0.3
      nose.closed = true
      nose.transform = Transformations.translation(x: 0, y: 2.2, z: 0.3) *
        Transformations.rotation_x(radians: Math::PI / 2) *
        Transformations.scaling(x: 0.08, y: 1, z: 0.08)
      nose.material.color = Color.new(r: 1, g: 0.5, b: 0)
      nose.material.specular = 0.1

      # Snowman group
      snowman = Group.new
      snowman.add_child(snowman_bottom)
      snowman.add_child(snowman_middle)
      snowman.add_child(snowman_head)
      snowman.add_child(nose)
      snowman.transform = Transformations.translation(x: 2, y: 0, z: -1)

      # Create a hexagon group (6 spheres in a ring)
      hexagon = Group.new
      6.times do |i|
        sphere = Sphere.new
        angle = i * Math::PI / 3
        sphere.transform = Transformations.translation(x: Math.cos(angle), y: 0.3, z: Math.sin(angle)) *
          Transformations.scaling(x: 0.3, y: 0.3, z: 0.3)

        # Color based on position
        hue = i / 6.0
        sphere.material.color = Color.new(
          r: (Math.sin(hue * Math::PI * 2) + 1) / 2,
          g: (Math.sin((hue + 0.33) * Math::PI * 2) + 1) / 2,
          b: (Math.sin((hue + 0.67) * Math::PI * 2) + 1) / 2
        )
        sphere.material.specular = 0.6
        sphere.material.reflective = 0.2

        hexagon.add_child(sphere)
      end
      hexagon.transform = Transformations.translation(x: 0, y: 0, z: 3)

      # Create world and add all objects
      world = World.new
      world.light = light
      world.objects << floor
      world.objects << tree
      world.objects << snowman
      world.objects << hexagon

      # Render the scene
      puts "Rendering 800x600 scene (this will take a few minutes)..."
      start_time = Time.now
      canvas = camera.render(world)
      end_time = Time.now
      elapsed = end_time - start_time

      puts "Rendering completed in #{elapsed.round(2)} seconds"
      puts "Writing to chapter13.ppm..."

      File.write("chapter13.ppm", canvas.to_ppm)
      puts "Done! Output saved to chapter13.ppm"
      puts "Scene demonstrates:"
      puts "  - Hierarchical groups (tree with trunk and foliage groups)"
      puts "  - Compound objects (snowman made from spheres and cylinder)"
      puts "  - Group transformations cascading to children"
      puts "  - Multiple grouped objects in a single scene"
    end
  end
end

File: python/examples/chapter13.py
"""Chapter 13: Groups — hierarchical scene composition."""

from __future__ import annotations

import math
import os

from rayz.camera import Camera
from rayz.color import Color
from rayz.cylinder import Cylinder
from rayz.group import Group
from rayz.pattern import checkers_pattern
from rayz.plane import Plane
from rayz.point_light import PointLight
from rayz.sphere import Sphere
from rayz.transformations import (
    rotation_x,
    scaling,
    translation,
    view_transform,
)
from rayz.tuple import Point, Vector
from rayz.world import World


def run() -> None:
    print("\n=== Chapter 13: Groups ===\n")

    world = World()
    world.light = PointLight(Point(-5, 10, -10), Color(1, 1, 1))

    floor = Plane()
    floor.material.pattern = checkers_pattern(Color(0.9, 0.9, 0.9), Color(0.1, 0.1, 0.1))
    floor.material.reflective = 0.1
    world.objects.append(floor)

    # Tree: trunk (cylinder) + foliage group (spheres)
    trunk = Cylinder()
    trunk.minimum = 0
    trunk.maximum = 1.5
    trunk.closed = True
    trunk.material.color = Color(0.4, 0.2, 0.1)
    trunk.material.specular = 0.1

    foliage = Group()
    for tx, ty, tz, sx in [(0, 2, 0, 0.8), (-0.5, 1.7, 0, 0.6), (0.5, 1.7, 0.2, 0.6), (0, 2.5, 0, 0.5)]:
        s = Sphere()
        s.set_transform(translation(tx, ty, tz) * scaling(sx, sx, sx))
        s.material.color = Color(0.1, 0.5 + sx * 0.3, 0.1)
        s.material.specular = 0.3
        foliage.add_child(s)

    tree = Group()
    tree.add_child(trunk)
    tree.add_child(foliage)
    tree.set_transform(translation(-2, 0, 1))
    world.objects.append(tree)

    # Snowman: 3 spheres + cylinder nose
    snowman = Group()
    for ty, sx in [(0.6, 0.6), (1.5, 0.45), (2.2, 0.3)]:
        s = Sphere()
        s.set_transform(translation(0, ty, 0) * scaling(sx, sx, sx))
        s.material.color = Color(1, 1, 1)
        s.material.specular = 0.9
        s.material.shininess = 300
        snowman.add_child(s)

    nose = Cylinder()
    nose.minimum = 0
    nose.maximum = 0.3
    nose.closed = True
    nose.set_transform(translation(0, 2.2, 0.3) * rotation_x(math.pi / 2) * scaling(0.08, 1, 0.08))
    nose.material.color = Color(1, 0.5, 0)
    nose.material.specular = 0.1
    snowman.add_child(nose)
    snowman.set_transform(translation(2, 0, -1))
    world.objects.append(snowman)

    # Hexagon: 6 colored spheres in a ring
    hexagon = Group()
    for i in range(6):
        angle = i * math.pi / 3
        s = Sphere()
        s.set_transform(translation(math.cos(angle), 0.3, math.sin(angle)) * scaling(0.3, 0.3, 0.3))
        h = i / 6.0
        s.material.color = Color(
            (math.sin(h * math.pi * 2) + 1) / 2,
            (math.sin((h + 0.33) * math.pi * 2) + 1) / 2,
            (math.sin((h + 0.67) * math.pi * 2) + 1) / 2,
        )
        s.material.specular = 0.6
        s.material.reflective = 0.2
        hexagon.add_child(s)
    hexagon.set_transform(translation(0, 0, 3))
    world.objects.append(hexagon)

    camera = Camera(200, 150, math.pi / 3)
    camera.transform = view_transform(Point(0, 3, -8), Point(0, 1, 0), Vector(0, 1, 0))

    print("Rendering 200x150 scene with groups (tree, snowman, hexagon)...")
    canvas = camera.render_parallel(world)

    out_path = os.path.join(os.path.dirname(__file__), "chapter13.ppm")
    print(f"Writing {out_path}...", end="", flush=True)
    with open(out_path, "w") as f:
        f.write(canvas.to_ppm())
    print(" done.")
    print("Groups scene rendered.")
    print("\n" + "=" * 60 + "\n")


if __name__ == "__main__":
    run()
